### PR TITLE
builder/triton: Add support for Triton RBAC

### PR DESCRIPTION
**Pull request by stack72**
Fixes: #5737

Triton allows for subusers to be granted access to your account. This PR allows a user to be specified

Looking at the following triton template:

```
{
    "variables": {
        "image_version": "",
        "triton_account": "",
        "triton_user": "",
        "triton_key_id": "",
        "triton_key_material": ""
    },
    "builders": [{
        "type": "triton",
        "triton_account": "{{user `triton_account`}}",
        "triton_key_id": "{{user `triton_key_id`}}",
        "triton_user": "{{user `triton_user`}}",
        "triton_key_material": "{{user `triton_key_material`}}",
        "source_machine_package": "g4-highcpu-128M",
        "source_machine_image_filter": {
          "name": "ubuntu-16.04",
          "most_recent": "true"
        },
	      "source_machine_firewall_enabled": false,
        "ssh_username": "root",
        "image_version": "{{user `image_version`}}",
        "image_name": "teamcity-server"
    }],
    "provisioners": [
        {
            "type": "shell",
            "start_retry_timeout": "10m",
            "inline": [
                "sudo apt-get update -y",
                "sudo apt-get install -y nginx"
            ]
        }
    ]
}
```

A packer build generates the following:

screen shot 2017-12-29 at 15 14 31

With the new addition to the code, I can see the following in my request headers when packer interacts with Triton:

screen shot 2017-12-29 at 15 17 26

Packer continues to work as normal when a user doesn't specify a triton_user

screen shot 2017-12-29 at 15 32 05

screen shot 2017-12-29 at 15 24 57

### CODE REVIEW

**Comment by rickard-von-essen** [comment]
Could you split your commit in two, one for the updated vendored lib and one with your enhancement?

**Comment by stack72** [comment]
Hi @rickard-von-essen 

I don't see what splitting the commits here will do - the dep update is only happening because of the new feature. The dep update itself is a breaking change as we are changing to an input struct so, to me, they seem coupled

Paul

**Comment by rickard-von-essen** [comment]
1) It makes reviewing simpler. 
2) If want to merge this after we allready merged another update of the same lib it's simpler to drop one full commit.
3) They comes from different authors, you and an upstream library.

**Comment by stack72** [comment]
There won't be an update of the same lib - I maintain triton-go as I work for Joyent

The PR is a change to these 2 files:

* builder/triton/access_config.go
* website/source/docs/builders/triton.html.md

I don't see how separate commits will make it easier to review

**Comment by rickard-von-essen** [comment]
Your PR changes 21 and I review alot of changes on my mobile phone so split the into 19 + 2 simplifies it for me (the reviewer).

**Comment by stack72** [comment]
So your reasoning for me to split a PR (that in my eyes shouldn't be separate commits) is to make it easier for you because you review things on your phone - seriously?

File: builder/triton/access_config.go [hunk 1]
```diff
@@ -94,8 +100,15 @@ func (c *AccessConfig) createPrivateKeySigner() (authentication.Signer, error) {
...
 	}
 
+	input := authentication.PrivateKeySignerInput{
+		KeyID:              c.KeyID,
+		AccountName:        c.Account,
+		Username:           c.Username,
+		PrivateKeyMaterial: privateKeyMaterial,
+	}
+
 	// Create signer
```
Comment (sean-) [inline builder/triton/access_config.go]: We can nuke this comment, this is inherent in the code itself.

File: website/source/docs/builders/triton.html.md [hunk 2]
```diff
@@ -92,6 +92,9 @@ builder.
     of `triton_key_id` is stored. For example `/home/soandso/.ssh/id_rsa`. If
     this is not specified, the SSH agent is used to sign requests with the
     `triton_key_id` specified.
+    
+-   `triton_user` (string) - The username of a user who has access to your triton
```
Comment (sean-) [inline website/source/docs/builders/triton.html.md]: Capital `T` in Triton.

File: website/source/docs/builders/triton.html.md [hunk 3]
```diff
@@ -149,7 +152,7 @@ builder.
 
 ## Basic Example
 
-Below is a minimal example to create an joyent-brand image on the Joyent public
+Below is a minimal example to create a joyent-brand image on the Joyent public
```
Comment (sean-) [inline website/source/docs/builders/triton.html.md]: I'm not sure what a "joyent-brand" is.  This is an Illumos Zone-based image, or an "infrastructure VM."  This nit predates this PR but now seems like a good time to fix this.

**Comment by rickard-von-essen** [comment]
@jen20 👍

**Pull request merged**